### PR TITLE
Save Img cls models with OSS flavor

### PR DESCRIPTION
**Pull request by MadhuM02**
Saving Img cls models with OSS flavor

ref runs: https://ml.azure.com/experiments/id/69a916d8-6a8d-42cc-9145-ec05e2228892?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/madhumaddi-rg/providers/Microsoft.MachineLearningServices/workspaces/madhumaddi-hd-onboard&tid=72f988bf-86f1-41af-91ab-2d7cd011db47

image

### CODE REVIEW

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py [hunk 1]
```diff
@@ -391,6 +394,25 @@ def get_model_signature(self):
...
+        """
+        ids = list(config.id2label.keys())
+        ids.sort()
+        if max(ids) != len(ids)-1:
+            id2label = {}
+            for idx, id in enumerate(ids):
+                id2label[idx] = config.id2label[id]
+            config.id2label = id2label
+            self.config = config
+            logger.warning("config loaded with modified id2label as there are some missing keys.")
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:413]: Would be good to call out, which ones were missing or added by us here/

**Review by rjaincc — APPROVED** [review]
Approving with an enhancement suggestion

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py (hunk 1 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py]: done, thanks

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py [hunk 2]
```diff
@@ -391,6 +394,26 @@ def get_model_signature(self):
...
 
+    def _robust_load_config(self, config) -> transformers.PretrainedConfig:
+        """Check and modify if config has missing indices.
+
+        :param config: Config Object from transformers
+        :type config: PretrainedConfig
+        :return: Config Object
+        :rtype: PretrainedConfig
+        """
+        ids = list(config.id2label.keys())
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:405]: id2label will be present for all models? Could we try catch or check if id2label exist before proceeding?

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py [hunk 3]
```diff
@@ -414,10 +438,21 @@ def save_as_mlflow(self):
...
 
-        return super()._save(
-            code_paths=[VisionMLflowConvertor.PREDICT_FILE_PATH, VisionMLflowConvertor.VISION_UTILS_FILE_PATH],
-            segregate=True,
-        )
+        if self._model_flavor == "OSS":
+            vision_model = transformers.pipeline(task=self._task, model=str(self._model_dir))
+            vision_model.model.config = config
+            image_path = str(Path(__file__).parent.parent / "common" / "test_image.jpg")
+            self._signatures = infer_signature(
```
Comment (novaturient95) [inline assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:445]: possible to infer without adding a test img?

**Comment by novaturient95** [comment]
not an issue with this PR but there is a LICENSE.txt that got added which is a dummy one in the model files: 

image


Is there anyway to not contain this one?

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py (hunk 2 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:405]: Good point Ayush. For saving it in transformers oss flavor, inference happens with id2label and its failing if the id2label is incorrectly formed. If it doesnt have id2label then the porting itself is not possible

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py (hunk 3 above)
Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:445]: We could do it like this:
    buffered = BytesIO()
    im = torch.rand((256, 256, 3))
    image = Image.fromarray(im.numpy().astype('uint8'), 'RGB')
    image.save(buffered, format="JPEG")
    img_str = base64.b64encode(buffered.getvalue()).decode("utf-8")
    return img_str
    
    But have to make sure bytestio and base64 packages are there. Would it be better to add these packages and get a mock_image instead??

Comment (MadhuM02) [reply assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:462]: Fixed this, can you please take a look?
ref run: https://ml.azure.com/experiments/id/69a916d8-6a8d-42cc-9145-ec05e2228892/runs/f922069b-ba70-4b7a-9aec-dd0e5f42e254?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/madhumaddi-rg/providers/Microsoft.MachineLearningServices/workspaces/madhumaddi-hd-onboard&tid=72f988bf-86f1-41af-91ab-2d7cd011db47

**Review by rjaincc — APPROVED** [review]

File: assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py [hunk 4]
```diff
@@ -215,7 +220,9 @@ def _save_in_oss_flavor(self, model, metadata, conda_env, code_paths, input_exam
 
         model_pipeline = transformers.pipeline(task=self._task, model=model,
                                                trust_remote_code=trust_remote_code_val)
-
+        if hasattr(self, "config"):
+            # Have to update the config as some vision models has problems in hf registry.
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/transformers/convertors.py:224]:
```suggestion
            # Have to update the config as some vision models have problems in hf registry.
```

**Review by rjaincc — APPROVED** [review]

**Review by rjaincc — APPROVED** [review]

**Pull request merged**